src/OpenLoco/src/GameCommands/Track/RemoveTrack.cpp

```diff
@@ -0,0 +1,295 @@
+        elTrack = getElTrackAt(args, flags, args.pos, args.index);
+
+        // 0x0049C9F0
+        if (elTrack->hasStationElement())
+        {
+            GameCommands::TrainStationRemovalArgs tsArgs;
+            tsArgs.pos = args.pos;
+            tsArgs.rotation = args.rotation;
+            tsArgs.trackId = args.trackId;
+            tsArgs.index = args.index;
+            tsArgs.type = args.trackObjectId;
+
+            if (auto cost = GameCommands::doCommand(tsArgs, flags); cost != FAILURE)
+            {
+                totalRemovalCost += cost;
+            }
+            else
+            {
+                return FAILURE;
+            }
+        }
+
+        // 0x0049CA3B
+        // Fetch track element again; station removal above might have invalidated the pointer.
+        elTrack = getElTrackAt(args, flags, args.pos, args.index);
+
+        // 0x0049CAC4
+        const auto trackPieces = World::TrackData::getTrackPiece(args.trackId);
+        auto& trackPiece = trackPieces[args.index];
+
+        const auto trackStart = args.pos - World::Pos3{ Math::Vector::rotate(World::Pos2{ trackPiece.x, trackPiece.y }, args.rotation), trackPiece.z };
+
+        // NB: moved out of the loop below (was at 0x0049CC1B)
+        const currency32_t pieceRemovalCost = trackRemoveCost(args, trackPieces[0], trackStart, flags);
+
+        bool trackHadBridge = false; // 0x0113605B
+        int8_t trackBridgeId = -1;   // 0x0113605C
+
+        // 0x0049CB27
+        for (auto& piece : trackPieces)
+        {
+            const auto trackLoc = trackStart + World::Pos3{ Math::Vector::rotate(World::Pos2{ piece.x, piece.y }, args.rotation), piece.z };
+
+            if (!(flags & Flags::aiAllocated))
+            {
+                World::TileManager::mapInvalidateTileFull(trackLoc); // shift piece.z?
+            }
+
+            // 0x0049CB84
+            auto* pieceElTrack = getElTrackAt(args, flags, trackLoc, piece.index);
+            if (pieceElTrack == nullptr)
+            {
+                return FAILURE;
+            }
+
+            // 0x0049CC02
+            if (pieceElTrack->hasBridge())
+            {
+                trackHadBridge = true;
+                trackBridgeId = pieceElTrack->bridge();
+            }
+
+            if (!(flags & Flags::apply))
+            {
+                continue;
+            }
+
+            auto baseHeight = pieceElTrack->baseHeight();
+            World::TileManager::removeElement(*reinterpret_cast<World::TileElement*>(pieceElTrack));
+            setLevelCrossingFlags(World::Pos3{ trackLoc.x, trackLoc.y, baseHeight });
+        }
+
+        totalRemovalCost += pieceRemovalCost;
+
+        if (trackHadBridge)
+        {
+            const auto* bridgeObj = ObjectManager::get<BridgeObject>(trackBridgeId);
+            totalRemovalCost += Economy::getInflationAdjustedCost(bridgeObj->sellCostFactor, bridgeObj->costIndex, 10);
```

Comment:
you could move this code into the `trackRemoveCost` function. Then we don't need to keep around the `trackHadBridge` and the `trackBridgeId`

Reply:
Currently, we only pass the first track piece to `trackRemoveCost`. However, it's possible for only part of the segment/track piece to have a bridge. I therefore think it's best to leave it as-is for now.